keps/sig-apps/4563-evacuation-api/README.md

```diff
@@ -0,0 +1,1194 @@
+	// progress by updating the .status.evacuationProgressTimestamp.
+	// If the .status.evacuationProgressTimestamp is not updated within the duration of
+	// .spec.progressDeadlineSeconds, the evacuated pod will be evicted using the Eviction API.
+	//
+	// The minimum value is 60 (1m) and the maximum value is 43200 (12h).
+	// The value must be greater than or equal to .spec.acceptDeadlineSeconds.
+	// This field is required and immutable.
+	ProgressDeadlineSeconds int32 `json:"progressDeadlineSeconds" protobuf:"varint,3,opt,name=progressDeadlineSeconds"`
+}
+
+// PodReference contains enough information to locate the referenced pod.
+type PodReference struct {
+	// Name of the pod.
+	// This field is required.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// Namespace of the pod.
+	// This field is required.
+	Namespace string `json:"namespace" protobuf:"bytes,2,opt,name=namespace"`
+	// UID of the pod.
+	// This field is required.
+	UID string `json:"uid" protobuf:"bytes,3,opt,name=uid"`
+}
+
+// EvacuationStatus represents the most recently observed status of the evacuation.
+type EvacuationStatus struct {
+	// EvacuatorRef references the evacuator of the current evacuation. 
+	// This field should be set when Accepted field is set.
+	// +optional
+	EvacuatorRef *EvacuatorReference `json:"evacuatorRef,omitempty" protobuf:"bytes,1,opt,name=evacuatorRef"`
+
+	// If set to true, EvacuationProgressTimestamp should also be set, and then updated periodically.
+	// If set to false, the pod will be evicted immediately.
+	// This field cannot be changed once set.
+	// +optional
+	Accepted *bool `json:"accepted,omitempty" protobuf:"varint,2,opt,name=accepted"`
+
+	// EvacuationProgressTimestamp is the time at which the evacuation was reported to be in progress by the evacuator.
+	// Cannot be set to the future time (after taking time skew into account).
+	// This field must be set when the Accepted field is set to true.
+	// +optional
+	EvacuationProgressTimestamp *metav1.Time `json:"evacuationProgressTimestamp,omitempty" protobuf:"bytes,3,opt,name=evacuationProgressTimestamp"`
+
+	// ExpectedEvacuationFinishTime is the time at which the evacuation is expected to end.
+	// May be empty if no estimate can be made.
+	// +optional
+	ExpectedEvacuationFinishTime *metav1.Time `json:"expectedEvacuationFinishTime,omitempty" protobuf:"bytes,4,opt,name=expectedEvacuationFinishTime"`
+	
+	// The number of unsuccessful attempts to evict the referenced pod, e.g. due to a PDB.
+	// This field is required.
+	FailedEvictionCounter int32 `json:"failedEvictionCounter" protobuf:"varint,5,opt,name=failedEvictionCounter"`
+	
+	// Message is a human readable message indicating details about the evacuation.
+	// This may be an empty string.
+	// +required
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MaxLength=32768
+	Message string `json:"message" protobuf:"bytes,6,opt,name=message"`
```

Comment:
Could we use Events for this instead? I think it's reasonable to log an Event at least once per hour about an ongoing attempted eviction, so there should always be at least one current Event available.

Reply:
Even if the persistency were not an issue, the events would be implemented by different evacuator implementations and we cannot force an API contract on them. So from the user's POV, they do not know which events to watch. 

The API ensures that there is only one place for the user to look at. But that does not mean events cannot be also used. Also conditions can be used to communicate evacuation state to humans or other components . I think the more observability options we have, the better.

Side note: there may be important information in the message why the evacuation is stuck.

Reply:
I have added a few points about the events to the KEP.

Reply:
> we cannot force an API contract on them

Assuming that someone wants to report an Event, we can document how: (eg) file an Event `regarding` the Pod, and use `related` to specify the associated Evacuation.

If a controller wants to file other Events they can, but it's easy for us to define what a controller is expected to do. AIUI `kubectl events` already supports useful ways to find events associated with the Pod or with the eviction request.

Reply:
Good idea, I have added this to the KEP so the users can find the events.